sstables/sstable_set.cc

```diff
@@ -61,94 +61,422 @@ std::ostream& operator<<(std::ostream& os, const sstables::sstable_run& run) {
     return os;
 }
 
-sstable_set::sstable_set(std::unique_ptr<sstable_set_impl> impl, schema_ptr s)
-        : _impl(std::move(impl))
-        , _schema(std::move(s))
-        , _all(make_lw_shared<sstable_list>()) {
-    // assumes impl contains no sstables
+versioned_sstable_set_data::versioned_sstable_set_data(std::unique_ptr<sstable_set_impl> impl) 
+    : _impl(std::move(impl)) {
 }
 
-sstable_set::sstable_set(const sstable_set& x)
-        : _impl(x._impl->clone())
-        , _schema(x._schema)
-        , _all(make_lw_shared<sstable_list>(*x._all))
-        , _all_runs(x._all_runs) {
+void versioned_sstable_set_data::insert(shared_sstable sst) {
+    if (_times_added.contains(sst)) {
+        _times_added.at(sst)++;
+        return;
+    }
+    try {
+        _impl->insert(sst);
```

Comment:
Unfortunately, this is also an O(n log n) operation in some cases. Those cases are for a partitioned_sstable_set, when a level 1 sstable is inserted or removed. A level 1 sstable covers 10% of the the token space managed the the node, and intersects with 10 L2 sstables and 100 L3 sstables. n log n for 100 is 700, so we need around 700 binary search tree operations for these cases.

Let's see the impact. If each operation takes 100ns (around 200 instructions, probably an over-estimate) then 700 operations take 70 usec. That's considerably lower than our 500 usec limit. If we extend it to L4 then we are at 10,000 operations or 1000 usec.

A table with a full-populated L3 overs 1000*160MB = 160GB per shard. An AWS i3 has 200GB per shard, so that's safe. I3en has 600 GB per shard, so it will spill into L4 (not fully populating it). I think we're safe though somewhat on the edge.